### PR TITLE
Add residue formula for summation

### LINKED ISSUE #21148: Infinite summation by residues
SymPy isn't able to compute the summation like 
```
x = Symbol('x')
Sum(1 / (x**2 + 1), (x, -oo, oo)).doit()
```

But if the summand is rational functions (where `f(x), g(x)` are polynomials) with degree constraint
Screen Shot 2021-03-22 at 21 29 51
(The reason of this degree constraint is related to the convergence of contour integral used in the derivation. The result can be different if `f(x), g(x)` are more than polynomials, but I defer such problems now because finding root can be more difficult for such general cases.)

It is possible to use residues to evaluate the summations as simple root sums if the original summand has no poles at integer values (And if it has, the summation would not be well defined)
However I would use the result that generalized up to the summation indices ranging across all integers minus the poles
The second formula is for alternating series.

Screen Shot 2021-03-23 at 09 22 05
Screen Shot 2021-03-23 at 09 22 10

It would also be possible to use this formula to give summation from 0 to infinity excluding the poles at integers (And finding even/oddness for polynomials is more evident than other expressions), such that it can be used to compute `Sum(1 / x**2, (x, 1, oo))`

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #21148

#### Brief description of what is fixed or changed

I have implemented the idea from #21148 which can solve some infinite summation problems that sympy can't solve.
Although I have limited for polynomial cases, I nontrivially extended it to handle summation over half interval for even functions, with taking account of extraneous evaluation for finite points.

However, there is still some more places left to improve the summation by residue formula to handle more generic classes of functions than polynomial, and also to handle shifted cases like `Sum(1 / (x**2 - 2*x + 2), (x, 1, oo))` which should be same as `Sum(1 / (x**2 + 1), (x, 0, oo))` but can't be recognized as an even function. (I tried using `decompose` but that doesn't work)

#### Release Notes

- concrete
  - Improved infinite summation capability by adding residue formula.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
Could `RootSum` be used to represent a not fully evaluated result in some of the cases that this doesn't handle?

**Comment by sylee957** [comment]
No, I'm unsure about whether that works because `residue` isn't a sympy object yet.

**Comment by oscarbenjamin** [comment]
> to handle shifted cases like `Sum(1 / (x**2 - 2*x + 2), (x, 1, oo))` which should be same as `Sum(1 / (x**2 + 1), (x, 0, oo))` but can't be recognized as an even function.

You could use `Poly.shift`. If the leading terms are `x^n + ax^(n-1)` for `n` even then a shift of `-a/n` would be needed to eliminate the first odd term:
```python
In [27]: Poly(x**4 - 28*x**3 + 297*x**2 - 1414*x + 2549, x).shift(S(28)/4)
Out[27]: Poly(x**4 + 3*x**2 + 1, x, domain='ZZ')
```

File: sympy/concrete/summations.py [hunk 1]
```diff
@@ -1331,6 +1340,177 @@ def eval_sum_hyper(f, i_a_b):
...
 
+def eval_sum_residue(f, i_a_b):
+    r"""Compute the infinite summation with residues
+
+    Notes
+    =====
+
+    If $f(n), g(n)$ are polynomials with $\deg(g(n)) - \deg(f(n)) >= 2$,
+    some infinite summations can be computed by the following residue
+    evaluations.
```
Comment (oscarbenjamin) [inline sympy/concrete/summations.py:1351]: Can you add references for the formulae used here?

**Comment by oscarbenjamin** [comment]
This looks good to me.

Perhaps the docstring could also explain which cases are currently handled but otherwise if you're done then +1 to merge.

**Pull request merged**